### ISSUE TITLE
[DI] Deprecate case insentivity of service identifiers

### ISSUE THREAD

**Issue by nicolas-grekas** [symfony/symfony#21193]
| Q                | A
| ---------------- | -----
| RFC?             | yes
| Symfony version  | 3.3

Case insensitivity blocks some optimizations in the dumped container, slows down getting some services (especially now that we're going to use class names as ids) and adds code complexity.
We'd better make them case sensitive in 4.0, thus deprecate before.

**Comment by fabpot** [comment]
I agree

**Comment by mvrhov** [comment]
So the recommended way of getting service by name will from 4.00 be `ClassName:class` as the IDE will help greatly there will be less typing however overriding/decorating the services will be annoying as the name won't represent the actual class anymore but the name itself will suggest that. ATM the service name could be seen as some kind of alias.

**Comment by stof** [comment]
@mvrhov we don't enforce using class names as id (it would break many use cases, as being able to replace an implementation is a great use case).
But we are making the ``id`` optional, using the class name in this case, to keep things simpler for newcomers.

**Comment by mvrhov** [comment]
Then the above doesn't make much sense (ok, it does, if the service name is class name). AFAIR the we switched once before to CI ones.

**Comment by stof** [comment]
I don't understand your answer

**Comment by chalasr** [comment]
Would doing the same for parameters be relevant?

**Comment by nicolas-grekas** [comment]
@chalasr I don't think so. I mean, deprecating for ids has a *first* technical reason, which is doing more inlining, thus perf. For params, unless you can envision one, there is no such gain.
